tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/op/core/NN.java

```diff
@@ -0,0 +1,379 @@
+   * into `labels`, pass label tensors through `tf.stop_gradient` before feeding it to this
+   * function.
+   *
+   * @param scope current scope
+   * @param labels Each vector along the class dimension should hold a valid probability
+   *     distribution e.g. for the case in which labels are of shape `[batch_size, num_classes]`,
+   *     each row of `labels[i]` must be a valid probability distribution.
+   * @param logits Per-label activations, typically a linear output. These activation energies are
+   *     interpreted as unnormalized log probabilities.
+   * @param axis The class dimension. -1 is the last dimension.
+   * @param <U> the data type of the logits
+   * @param <T> the number type of the operands
+   * @return the softmax cross entropy loss. Its type is the same as `logits` and its shape is the
+   *     same as `labels` except that it does not have the last dimension of `labels`.
+   */
+  @Endpoint(name = "softmaxCrossEntropyWithLogits")
+  public static <U extends TType, T extends TNumber> Operand<T> softmaxCrossEntropyWithLogits(
+      Scope scope, Operand<T> labels, Operand<U> logits, int axis) {
+    axis = axis % logits.asOutput().shape().numDimensions();
+    if (axis < 0) {
+      axis += logits.asOutput().shape().numDimensions();
+    }
+
+    Operand precise_logits =
+        logits; // cannot use generics cause logits of bool gets cast to TFloat32
+
+    boolean convertToFloat32 =
+        logits.asOutput().dataType() == TFloat16.DTYPE
+            || logits.asOutput().dataType() == TBfloat16.DTYPE;
+    if (convertToFloat32) {
+      precise_logits = Cast.create(scope, logits, TFloat32.DTYPE);
+    }
+    /* cannot use generics on DataType because precis_logits may have been cast. */
+    DataType dtype = precise_logits.asOutput().dataType();
+    labels = Cast.create(scope, labels, dtype);
+    Operand<TInt64> inputRank =
+        Cast.create(scope, Rank.create(scope, precise_logits), TInt64.DTYPE);
+    Shape shape = logits.asOutput().shape();
+
+    // Move the dim to the end if dim is not the last dimension.
+    if (axis != -1 && axis != precise_logits.asOutput().shape().numDimensions() - 1) {
+      precise_logits = moveDimToEnd(scope, precise_logits, axis, inputRank);
+      labels = moveDimToEnd(scope, labels, axis, inputRank);
+    }
+
+    Shape inputShape = precise_logits.asOutput().shape();
+    precise_logits = flattenOuterDims(scope, precise_logits);
+    labels = flattenOuterDims(scope, labels);
+    SoftmaxCrossEntropyWithLogits<T> smax =
+        SoftmaxCrossEntropyWithLogits.create(scope, precise_logits, labels);
+    /* cannot use generic on cost, because cost may be recast later. */
+    Operand cost = smax.loss();
+    Operand<TInt64> outputShape =
+        Slice.create(
+            scope,
+            Constant.vectorOf(scope, inputShape.asArray()),
+            Constant.vectorOf(scope, new long[] {0}),
+            Constant.vectorOf(scope, new long[] {inputShape.numDimensions() - 1}));
```

Comment:
Shortcut: you can [`Constant.tensorOf(scope, inputShape)`](https://github.com/tensorflow/java/blob/86840e7766f1f321267c5bf2f2333eefcd659251/tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/op/core/Constant.java#L1255) directly for the first, and [`Constant.arrayOf(scope, num)`](https://github.com/tensorflow/java/blob/86840e7766f1f321267c5bf2f2333eefcd659251/tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/op/core/Constant.java#L108) for the others